test/test_loopy.py

```diff
@@ -842,6 +842,81 @@ def test_auto_test_zero_warmup_rounds(ctx_factory):
             warmup_rounds=0)
 
 
+def test_auto_test_allows_int_shape_dims_represented_as_float_vars(ctx_factory):
+    # Arrays with integer shape dims should be allowed even if
+    # the variable resulting from pymbolic.evaluate has type float
+    # (e.g., dim shapes evaluating to float 64.0 should be allowed)
+
+    ctx = ctx_factory()
+    queue = cl.CommandQueue(ctx)
+
+    # {{{ Arrays with integer shape dims should be allowed
+
+    # {{{ Create knl args
+
+    dim_size = 63
+    n = dim_size - 2
+
+    v = cl.array.empty(queue, (int(dim_size/2*2),), dtype=np.float32)
+    cl.clrandom.fill_rand(v)
+
+    knl_arg_dict = {"v": v, "dim_size": dim_size, "n": n}
+
+    # }}}
+
+    # {{{ Make kernel with integer v shape
+
+    knl = lp.make_kernel(
+        "[n] -> { [i] : 1 <= i < n }",
+        "v[i] = 3.14",
+        [
+            lp.GlobalArg("v", dtype=np.float32, shape=("dim_size/2*2")),
+            lp.ValueArg("n", dtype=np.int32),
+            lp.ValueArg("dim_size", dtype=np.int32),
+        ],
+        )
+
+    # }}}
+
+    # This should work because v's shape is an integer
+    # (even though it has type float when evaluated):
+    ref_knl = knl
+    lp.auto_test_vs_ref(ref_knl, ctx, knl, parameters=knl_arg_dict)
+
+    # }}}
+
+    # {{{ Arrays with non-integer shape dims should not be allowed
+
+    # {{{ Make kernel with non-integer v shape
+
+    v = cl.array.empty(queue, (int(dim_size/5*6),), dtype=np.float32)
+    cl.clrandom.fill_rand(v)
+
+    knl_arg_dict = {"v": v, "dim_size": dim_size, "n": n}
+
+    knl = lp.make_kernel(
+        "[n] -> { [i] : 1 <= i < n }",
+        "v[i] = 3.14",
+        [
+            lp.GlobalArg("v", dtype=np.float32, shape=("dim_size/5*6")),
```

Comment:
```suggestion
            lp.GlobalArg("v", dtype=np.float32, shape=("dim_size//5*6")),
```
I kind of disagree with the premise here. I think the issue is resolved better by using Python's integer division operator (`//`, [docs](https://docs.python.org/3/reference/expressions.html#binary-arithmetic-operations)), which pymbolic supports. What's the source of the divisions you're encountering?

Reply:
@inducer Then maybe there's a Fortran parsing problem? [Here's an example snippet that demonstrates the error when `auto-test-vs-ref` is called on the (unmodified) `blts.f90` kernel](https://gitlab.tiker.net/jdsteve2/2021-thesis-evaluation/-/blob/master/blts-auto-test-fail.py)

To run it, the only other file you need is [blts.f90](https://gitlab.tiker.net/jdsteve2/2021-thesis-evaluation/-/blob/master/blts.f90).

Reply:
Thanks for coming up with this reproducer!

Yeah, I think fixing the Fortran parser is the better direction. I'll see what I can come up with. Unfortunately, [Fortran integer division semantics](http://www.mrao.cam.ac.uk/~pa/f90Notes/HTMLNotesnode87.html) are different from Python for negative integers. (The Python way makes mathematical sense, the Fortran one doesn't. C uses the same brain-dead convention as Fortran.) I hope you don't need this to be right for negative integers... otherwise this'll be a nice bit of work.

Reply:
Btw, do you know what the license of the NAS parallel benchmarks is? I tried finding out, and it wasn't obvious. Their web page also doesn't seem to be responding for me.

Reply:
#364

Reply:
@inducer Thanks for your help with this! I don't believe negative integers will be an issue for me. Regarding the copyright question, not sure if this is definitive, but there are several techincal reports related to the benchmarks [here on the NASA Technical Reports Server](https://ntrs.nasa.gov/search?q=nas%20parallel%20benchmarks) that say "Copyright: Work of the US Gov. Public Use Permitted." The NPB website is still down, and I haven't found anything else besides an old git repo that uses the benchmarks and refers to "Copyright 1991-2014 NASA Advanced Supercomputing Division".

Reply:
The license concern related to using `blts` as part of a test. I ended up writing a smaller, self-contained test instead, so it's no longer a worry. Thanks for checking though!